ingest/src/test/java/com/microsoft/azure/kusto/ingest/IngestClientImplTest.java

```diff
@@ -107,12 +116,127 @@ void ingestFromStream() {
+
+        assertThrows(IngestionClientException.class,
+                () -> ingestClientSpy.ingestFromResultSet(resultSetSourceInfo, ingestionProperties));
+    }
+
+    @Test
+    void ingestFromResultSet_FileIngest_IngestionClientException() throws IngestionClientException, IngestionServiceException, SQLException {
+        IngestClient ingestClient = new IngestClientImpl(resourceManagerMock, azureStorageHelperMock);
+        // we need a spy to intercept the call to ingestFromFile so it wouldn't be called
+        IngestClient ingestClientSpy = spy(ingestClient);
+
+        IngestionClientException ingestionClientException = new IngestionClientException("Client exception in ingestFromFile");
+        doThrow(ingestionClientException).when(ingestClientSpy).ingestFromFile(any(), any());
+
+        ResultSet resultSet = getSampleResultSet();
+        ResultSetSourceInfo resultSetSourceInfo = new ResultSetSourceInfo(resultSet);
+
+        assertThrows(IngestionClientException.class,
+                () -> ingestClientSpy.ingestFromResultSet(resultSetSourceInfo, ingestionProperties));
+    }
+
+    @Test
+    void ingestFromResultSet_FileIngest_IngestionServiceException() throws IngestionClientException, IngestionServiceException, SQLException {
+        IngestClient ingestClient = new IngestClientImpl(resourceManagerMock, azureStorageHelperMock);
+        // we need a spy to intercept the call to ingestFromFile so it wouldn't be called
+        IngestClient ingestClientSpy = spy(ingestClient);
+
+        IngestionServiceException ingestionServiceException = new IngestionServiceException("Service exception in ingestFromFile");
+        doThrow(ingestionServiceException).when(ingestClientSpy).ingestFromFile(any(), any());
+
+        ResultSet resultSet = getSampleResultSet();
+        ResultSetSourceInfo resultSetSourceInfo = new ResultSetSourceInfo(resultSet);
+
+        assertThrows(IngestionServiceException.class,
+                () -> ingestClientSpy.ingestFromResultSet(resultSetSourceInfo, ingestionProperties));
+    }
+
+    @Test
+    void resultSetToCsv_Success() throws SQLException, IngestionClientException {
+        IngestClientImpl ingestClient = new IngestClientImpl(resourceManagerMock, azureStorageHelperMock);
+        ResultSet resultSet = getSampleResultSet();
+        StringWriter stringWriter = new StringWriter();
+
+        long numberOfCharsActual = ingestClient.resultSetToCsv(resultSet, stringWriter, false);
+
+        final String expected = System.getProperty("line.separator").equals("\n") ?
+                "\"1\",\"leo\"\n\"2\",\"yui\"\n" :
+                "\"1\",\"leo\"\r\n\"2\",\"yui\"\r\n";
+
+        assertEquals(expected, stringWriter.toString()); // check the string values
+        assertEquals(expected.length(), numberOfCharsActual); // check the returned length
+    }
+
+    @Test
+    void resultSetToCsv_ResultSetClosed_Exception() throws IngestionClientException, IngestionServiceException, SQLException {
```

Comment:
Is ResultSetClosed the only case when an exception is closed?

Reply:
Added another one for the writer